File: android/app/src/main/kotlin/com/nkkuma/shogi_movie_flutter/shogi_movie_flutter/domain/Board.kt
package com.nkkuma.shogi_movie_flutter.shogi_movie_flutter.domain

data class Board(
    var blackPiece: Map<Piece, Int> = mapOf(),
    var whitePiece: Map<Piece, Int> = mapOf(),
    var board: MutableList<MutableList<Piece>> = mutableListOf(
        mutableListOf(Piece.VKYO, Piece.VKEI, Piece.VGIN, Piece.VKIN, Piece.VOU, Piece.VKIN, Piece.VGIN, Piece.VKEI, Piece.VKYO),
        mutableListOf(Piece.NONE, Piece.VHISYA, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.VKAKU, Piece.NONE),
        mutableListOf(Piece.VFU, Piece.VFU, Piece.VFU, Piece.VFU, Piece.VFU, Piece.VFU, Piece.VFU, Piece.VFU, Piece.VFU),
        mutableListOf(Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE),
        mutableListOf(Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE),
        mutableListOf(Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE),
        mutableListOf(Piece.FU, Piece.FU, Piece.FU, Piece.FU, Piece.FU, Piece.FU, Piece.FU, Piece.FU, Piece.FU),
        mutableListOf(Piece.NONE, Piece.KAKU, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.HISYA, Piece.NONE),
        mutableListOf(Piece.KYO, Piece.KEI, Piece.GIN, Piece.KIN, Piece.GYOKU, Piece.KIN, Piece.GIN, Piece.KEI, Piece.KYO)
    )
) {
    fun clearBoard() = run { board = MutableList(9) { MutableList(9) { Piece.NONE } } }

    fun toSfen(): String {
        val sfenArray = arrayListOf<String>()
        for (line in board) {
            var spaceNumber = 0
            val mergedSfen = StringBuilder()
            for (place in line) {
                if (place != Piece.NONE) {
                    if (spaceNumber > 0) {
                        mergedSfen.append(spaceNumber.toString())
                        spaceNumber = 0
                    }
                    mergedSfen.append(place.sfen)
                }
                else {
                    spaceNumber += 1
                }
            }
            if (spaceNumber > 0) {
                mergedSfen.append(spaceNumber.toString())
            }
            sfenArray.add(mergedSfen.toString())
        }
        return sfenArray.joinToString(separator = "/")
    }

    fun fromSfen(sfen: String): Board {
        val board = MutableList(9) { MutableList(9) { Piece.NONE } }
        var index = 0
        var promote = false
        for (char in sfen) {
            if (char == '/') {
                index += 1
                promote = false
            }
            else if (char == '+') {
                promote = true
            }
            else if (char == ' ') {
                // it starts mochigoma and kif. currently these are ignored.
                break
            }
            else if (!char.isDigit()) {
                val pieceStr = if (promote) "+$char" else char.toString()
                board[index/10][index%10] = Piece.values().first { it.sfen == pieceStr }
                promote = false
                index += 1
            }
            else {
                for (i in 1..Character.getNumericValue(char)) {
                    board[index/10][index%10] = Piece.NONE
                    promote = false
                    index += 1
                }
            }
        }
        return Board(mapOf(), mapOf(), board)
    }
}

File: lib/domain/board.dart
import 'package:shogi_movie_flutter/domain/piece.dart';
import 'package:shogi_movie_flutter/domain/piece_list.dart';

class Board{
  Map<Piece, int> blackPiece = {};
  Map<Piece, int> whitePiece = {};
  List<List<Piece>> board = [
    [Piece.VKYO, Piece.VKEI, Piece.VGIN, Piece.VKIN, Piece.VOU, Piece.VKIN, Piece.VGIN, Piece.VKEI, Piece.VKYO],
    [Piece.NONE, Piece.VHISYA, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.VKAKU, Piece.NONE],
    [Piece.VFU, Piece.VFU, Piece.VFU, Piece.VFU, Piece.VFU, Piece.VFU, Piece.VFU, Piece.VFU, Piece.VFU],
    [Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE],
    [Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE],
    [Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE],
    [Piece.FU, Piece.FU, Piece.FU, Piece.FU, Piece.FU, Piece.FU, Piece.FU, Piece.FU, Piece.FU],
    [Piece.NONE, Piece.KAKU, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.NONE, Piece.HISYA, Piece.NONE],
    [Piece.KYO, Piece.KEI, Piece.GIN, Piece.KIN, Piece.GYOKU, Piece.KIN, Piece.GIN, Piece.KEI, Piece.KYO]
  ];
  
  void clearBoard() {
    board = List.generate(9, (_) => List.generate(9, (_) => Piece.NONE));
  }

  Board();

  Board.init(this.blackPiece, this.whitePiece, this.board);

  Board fromSfen(String sfen) {
    Map<Piece, int> blackPiece = {};
    Map<Piece, int> whitePiece = {};
    List<List<Piece>> board = List.generate(9, (_) => List.generate(9, (_) => Piece.NONE));
    int index = 0;
    bool promote = false;
    String sfenBoard = sfen.split(" ")[0];
    for (String char in sfenBoard.split("")) {
      if (char == "/") {
        promote = false;
        index += 1;
      }
      else if (char == "+") {
        promote = true;
      }
      else if (int.tryParse(char) == null) {
        Piece piece = Piece.values.firstWhere((element) => element.sfen == char);
        board[index~/10][index%10] = promote ? piece.promoted : piece;
        promote = false;
        index += 1;
      }
      else {
        for (int i = 1; i <= int.parse(char); i++) {
          board[index~/10][index%10] = Piece.NONE;
          promote = false;
          index += 1;
        }
      }
    }

    if (sfen.split(" ").length > 1) {
      // ignore turn (w or b)

      // blackPiece and whitePiece
      int num = 1;
      String pieces = sfen.split(" ")[2];
      for (String char in pieces.split("")) {
        if (int.tryParse(char) == null) {
          Piece piece = Piece.values.firstWhere((element) => element.sfen == char);
          firstPieceList.contains(piece) ? blackPiece[piece] = num : whitePiece[piece] = num;
          num = 1;
        }
        else {
          num = int.parse(char);
        }
      }

      // ignore tesuu and kifu
    }

    return Board.init(blackPiece, whitePiece, board);
  }

  String toSfen() {
    List<String> sfenArray = [];
    for (List<Piece> line in board) {
      int spaceNumber = 0;
      String mergedSfen = "";
      for (Piece place in line) {
        if (place != Piece.NONE) {
          if (spaceNumber > 0) {
            mergedSfen += spaceNumber.toString();
            spaceNumber = 0;
          }
          mergedSfen += place.sfen;
        }
        else {
          spaceNumber += 1;
        }
      }
      if (spaceNumber > 0) {
        mergedSfen += spaceNumber.toString();
      }
      sfenArray.add(mergedSfen);
    }
    String boardStr = sfenArray.join("/");
    if (blackPiece.isEmpty && whitePiece.isEmpty) return boardStr;
    String turn = "b";
    String pieces = "";
    blackPiece.forEach((key, value) {
      value == 1 ? pieces += key.sfen : pieces += value.toString() + key.sfen;
    });
    whitePiece.forEach((key, value) {
      value == 1 ? pieces += key.sfen : pieces += value.toString() + key.sfen;
    });
    String tesuu = "1";
    return boardStr + " " + turn + " " + pieces + " " + tesuu;
  }

  String toKif() {
    List<String> sfenArray = [];
    for (List<Piece> line in board) {
      int spaceNumber = 0;
      String mergedSfen = "";
      for (Piece place in line) {
        if (place != Piece.NONE) {
          if (spaceNumber > 0) {
            mergedSfen += spaceNumber.toString();
            spaceNumber = 0;
          }
          mergedSfen += place.sfen;
        }
        else {
          spaceNumber += 1;
        }
      }
      if (spaceNumber > 0) {
        mergedSfen += spaceNumber.toString();
      }
      sfenArray.add(mergedSfen);
    }
    String boardStr = sfenArray.join("/");
    if (blackPiece.isEmpty && whitePiece.isEmpty) return boardStr;
    String turn = "b";
    String pieces = "";
    blackPiece.forEach((key, value) {
      value == 1 ? pieces += key.sfen : pieces += value.toString() + key.sfen;
    });
    whitePiece.forEach((key, value) {
      value == 1 ? pieces += key.sfen : pieces += value.toString() + key.sfen;
    });
    String tesuu = "1";
    return boardStr + " " + turn + " " + pieces + " " + tesuu;
  }
}